File: reference-implementations/python/ledger/lots.py
"""
Minimal lots / cost basis relief for the Python reference canonical.

Faithful subset of src/portfolio/lots.ts for FIFO (expandable to LIFO/HIFO).

Reconstructs open lots and realized P&L purely from ledger entries that carry
LOT_TAGS on CUST: custody lines (as produced by trading.py reconcile helpers).

All arithmetic uses Money only. Deterministic. Fails closed on oversell.
"""

from dataclasses import dataclass
from decimal import Decimal
from typing import List, Optional, Dict, Any

from .money import Money
from .ledger import Ledger
from .trading import LOT_TAGS


@dataclass(frozen=True)
class Lot:
    id: str
    asset: str
    acquired_date: str
    origin_entry_id: str
    quantity: Money
    cost_basis: Money


@dataclass(frozen=True)
class RealizedDisposal:
    trade_id: str
    date: str
    asset: str
    quantity: Money
    proceeds: Money
    basis: Money
    gain: Money


@dataclass(frozen=True)
class ReliefResult:
    asset: str
    quote: str
    open_lots: List[Lot]
    realized: List[RealizedDisposal]
    total_realized: Money


def _events_for(ledger: Ledger, asset: str) -> Dict[str, Any]:
    A = asset.upper()
    events: List[Dict[str, Any]] = []
    quote: Optional[str] = None
    for entry in ledger.entries:
        for line in entry.lines:
            t = line.tags or {}
            if not t.get(LOT_TAGS["tradeId"]):
                continue
            if line.amount.currency != A:
                continue
            if not line.account.code.startswith("CUST:"):
                continue
            q = t.get(LOT_TAGS["quote"])
            if q:
                if quote and quote != q:
                    raise ValueError(f"Asset {A} mixed quotes {quote} vs {q}")
                quote = q
            role = t.get(LOT_TAGS["role"])
            if role == "acquire":
                basis_str = t.get(LOT_TAGS["costBasis"], "0")
                events.append({
                    "kind": "acquire",
                    "date": entry.effective_date,
                    "entry_id": entry.id,
                    "qty": line.amount,
                    "basis": Money.from_(basis_str, q or "USD"),
                })
            elif role == "dispose":
                proceeds_str = t.get(LOT_TAGS["proceeds"], "0")
                events.append({
                    "kind": "dispose",
                    "date": entry.effective_date,
                    "trade_id": t.get(LOT_TAGS["tradeId"]),
                    "qty": line.amount,
                    "proceeds": Money.from_(proceeds_str, q or "USD"),
                })
    return {"events": events, "quote": quote or "USD"}


def _pick_lot(lots: List[Lot], method: str = "FIFO") -> int:
    if not lots:
        return -1
    if method.upper() == "FIFO":
        return 0
    if method.upper() == "LIFO":
        return len(lots) - 1
    # HIFO minimal
    best = 0
    best_per = lots[0].cost_basis.to_decimal() / lots[0].quantity.to_decimal()
    for i in range(1, len(lots)):
        per = lots[i].cost_basis.to_decimal() / lots[i].quantity.to_decimal()
        if per > best_per or (per == best_per and lots[i].id < lots[best].id):
            best = i
            best_per = per
    return best


def relief_for(ledger: Ledger, asset: str, method: str = "FIFO") -> ReliefResult:
    A = asset.upper()
    data = _events_for(ledger, A)
    q = data["quote"]
    lots: List[Lot] = []
    realized: List[RealizedDisposal] = []
    total_realized = Money.zero(q)
    acquire_seq = 0

    for ev in data["events"]:
        if ev["kind"] == "acquire":
            lots.append(Lot(
                id=f"{ev['entry_id']}#{acquire_seq}",
                asset=A,
                acquired_date=ev["date"],
                origin_entry_id=ev["entry_id"],
                quantity=ev["qty"],
                cost_basis=ev["basis"],
            ))
            acquire_seq += 1
            continue

        # dispose
        remaining = ev["qty"].to_decimal()
        proceeds_total = ev["proceeds"]
        basis_sum = Money.zero(q)

        while remaining > 0:
            idx = _pick_lot(lots, method)
            if idx < 0:
                raise ValueError(f"Oversell of {A}: {ev['trade_id']}")
            lot = lots[idx]
            lot_qty = lot.quantity.to_decimal()
            take = min(remaining, lot_qty)
            full = take >= lot_qty

            if full:
                basis_take = lot.cost_basis
                lots.pop(idx)
            else:
                ratio = take / lot_qty
                basis_take = lot.cost_basis.mul(str(ratio))
                new_qty = lot.quantity.mul(str( (lot_qty - take) / lot_qty ))
                new_basis = lot.cost_basis.sub(basis_take)
                lots[idx] = Lot(lot.id, lot.asset, lot.acquired_date, lot.origin_entry_id, new_qty, new_basis)

            basis_sum = basis_sum.add(basis_take)
            remaining -= take

        gain = proceeds_total.sub(basis_sum)
        total_realized = total_realized.add(gain)
        realized.append(RealizedDisposal(
            trade_id=ev["trade_id"],
            date=ev["date"],
            asset=A,
            quantity=ev["qty"],
            proceeds=proceeds_total,
            basis=basis_sum,
            gain=gain,
        ))

    return ReliefResult(
        asset=A,
        quote=q,
        open_lots=lots,
        realized=realized,
        total_realized=total_realized,
    )


def build_lots(ledger: Ledger, asset: str, method: str = "FIFO") -> List[Lot]:
    return relief_for(ledger, asset, method).open_lots


def realized_pnl(ledger: Ledger, asset: str, method: str = "FIFO") -> ReliefResult:
    return relief_for(ledger, asset, method)

File: src/portfolio/lots.ts
import { Money, ROUND_HALF_UP } from '../core/money.js';
import { Ledger } from '../core/ledger.js';
import { LOT_TAGS } from '../trading/postings.js';


export type LotMethod = 'FIFO' | 'LIFO' | 'HIFO';

/** An open tax/cost lot: remaining quantity of an asset and its remaining basis. */
export interface Lot {
  readonly id: string;            // deterministic: derived from the acquiring entry
  readonly asset: string;
  readonly acquiredDate: string;
  readonly originEntryId: string;
  readonly quantity: Money;       // remaining, in asset units
  readonly costBasis: Money;      // remaining total basis, in quote currency
}

export interface RealizedDisposal {
  readonly tradeId: string;
  readonly date: string;
  readonly asset: string;
  readonly quantity: Money;       // disposed, in asset units
  readonly proceeds: Money;       // quote
  readonly basis: Money;          // quote
  readonly gain: Money;           // quote (negative = loss)
}

export interface ReliefResult {
  readonly asset: string;
  readonly quote: string;
  readonly openLots: Lot[];
  readonly realized: RealizedDisposal[];
  readonly totalRealized: Money;  // quote
}

interface Acquire { kind: 'acquire'; date: string; entryId: string; qty: Money; basis: Money; }
interface Dispose { kind: 'dispose'; date: string; tradeId: string; qty: Money; proceeds: Money; }

/** Pull acquisition/disposal events for an asset from custody lines, in ledger order. */
function eventsFor(ledger: Ledger, asset: string): { events: (Acquire | Dispose)[]; quote: string | undefined } {
  const A = asset.toUpperCase();
  const events: (Acquire | Dispose)[] = [];
  let quote: string | undefined;
  for (const entry of ledger.entries) {
    for (const line of entry.lines) {
      const t = line.tags;
      if (!t || !t[LOT_TAGS.tradeId]) continue;
      if (line.amount.currency !== A) continue;
      if (!line.account.code.startsWith('CUST:')) continue;
      const q = t[LOT_TAGS.quote];
      if (q) {
        if (quote && quote !== q) throw new Error(`Asset ${A} has mixed quote currencies (${quote} vs ${q})`);
        quote = q;
      }
      if (t[LOT_TAGS.role] === 'acquire') {
        events.push({
          kind: 'acquire', date: entry.effectiveDate, entryId: entry.id,
          qty: line.amount, basis: Money.from(t[LOT_TAGS.costBasis]!, q!),
        });
      } else if (t[LOT_TAGS.role] === 'dispose') {
        events.push({
          kind: 'dispose', date: entry.effectiveDate, tradeId: t[LOT_TAGS.tradeId]!,
          qty: line.amount, proceeds: Money.from(t[LOT_TAGS.proceeds]!, q!),
        });
      }
    }
  }
  return { events, quote };
}

/** Choose the next lot index to relieve under the method (lots kept in acquisition order). */
function pickLot(lots: Lot[], method: LotMethod): number {
  if (lots.length === 0) return -1;
  if (method === 'FIFO') return 0;
  if (method === 'LIFO') return lots.length - 1;
  // HIFO: highest cost-basis-per-unit first (tax-loss-harvesting style), deterministic tie-break by id.
  let best = 0;
  let bestPer = lots[0].costBasis.toDecimal().div(lots[0].quantity.toDecimal());
  for (let i = 1; i < lots.length; i++) {
    const per = lots[i].costBasis.toDecimal().div(lots[i].quantity.toDecimal());
    if (per.cmp(bestPer) > 0 || (per.cmp(bestPer) === 0 && lots[i].id < lots[best].id)) {
      best = i; bestPer = per;
    }
  }
  return best;
}

/**
 * Reconstruct open lots and realized P&L for one asset from the immutable ledger.
 * Pure, deterministic projection — no parallel state. Fails closed on oversell.
 */
export function reliefFor(ledger: Ledger, asset: string, method: LotMethod = 'FIFO'): ReliefResult {
  const A = asset.toUpperCase();
  const { events, quote } = eventsFor(ledger, A);
  const q = quote ?? 'USD';
  const quoteScale = Money.zero(q).scale;
  const lots: Lot[] = [];
  const realized: RealizedDisposal[] = [];
  let totalRealized = Money.zero(q);
  let acquireSeq = 0;

  for (const ev of events) {
    if (ev.kind === 'acquire') {
      lots.push({
        id: `${ev.entryId}#${acquireSeq++}`, asset: A,
        acquiredDate: ev.date, originEntryId: ev.entryId,
        quantity: ev.qty, costBasis: ev.basis,
      });
      continue;
    }
    // dispose: relieve qty across lots per method
    let remaining = ev.qty.toDecimal();
    const proceedsTotal = ev.proceeds.toDecimal();
    const qtyTotal = ev.qty.toDecimal();
    let basisSum = Money.zero(q);
    let proceedsAllocated = Money.zero(q);

    while (remaining.gt(0)) {
      const idx = pickLot(lots, method);
      if (idx < 0) {
        throw new Error(`Oversell of ${A}: disposal ${ev.tradeId} exceeds held quantity`);
      }
      const lot = lots[idx];
      const lotQty = lot.quantity.toDecimal();
      const take = remaining.gte(lotQty) ? lotQty : remaining;
      const full = take.gte(lotQty);

      // Basis for the slice: whole remaining basis if lot fully consumed, else proportional.
      const basisTake = full
        ? lot.costBasis
        : Money.from(lot.costBasis.toDecimal().mul(take).div(lotQty).toDecimalPlaces(quoteScale, ROUND_HALF_UP).toString(), q);
      // Proceeds for the slice: proportional, with the final slice taking the remainder
      // so disposal proceeds reconcile exactly to the tagged total.
      const isLastSlice = remaining.minus(take).lte(0);
      const proceedsTake = isLastSlice
        ? ev.proceeds.sub(proceedsAllocated)
        : Money.from(proceedsTotal.mul(take).div(qtyTotal).toDecimalPlaces(quoteScale, ROUND_HALF_UP).toString(), q);

      basisSum = basisSum.add(basisTake);
      proceedsAllocated = proceedsAllocated.add(proceedsTake);

      // shrink or remove the lot
      if (full) {
        lots.splice(idx, 1);
      } else {
        lots[idx] = {
          ...lot,
          quantity: Money.from(lotQty.minus(take).toString(), A),
          costBasis: lot.costBasis.sub(basisTake),
        };
      }
      remaining = remaining.minus(take);
    }

    const gain = ev.proceeds.sub(basisSum);
    realized.push({
      tradeId: ev.tradeId, date: ev.date, asset: A,
      quantity: ev.qty, proceeds: ev.proceeds, basis: basisSum, gain,
    });
    totalRealized = totalRealized.add(gain);
  }

  return { asset: A, quote: q, openLots: lots, realized, totalRealized };
}

/** Open lots for an asset (convenience). */
export function buildLots(ledger: Ledger, asset: string, method: LotMethod = 'FIFO'): Lot[] {
  return reliefFor(ledger, asset, method).openLots;
}
